### ISSUE TITLE
cleanup BatchingExecutor and execution contexts

### ISSUE THREAD

**Issue by patriknw** [akka/akka#26690]
A few things that I think we can cleanup:
* Why do we have `akka.dispatch.BatchingExecutor`? What is the difference between this and then one in Scala?
* `ActorSystem.internalCallingThreadExecutionContext` vs `akka.dispatch.ExecutionContexts.sameThreadExecutionContext`
* Use 2.13/2.12 specific source files instead of loading with reflection as was introduced as a minimal fix in https://github.com/akka/akka/pull/26683, at least 2.13 parasitic is public

**Comment by viktorklang** [comment]
1) The one in Scala is not public API
2) I'd remove `ActorSystem.internalCallingThreadExecutionContext` and have `akka.dispatch.ExecutionContexts.sameThreadExecutionContext` delegate to `ExecutionContext.parasitic` (or try to load the Future.InternalCallbackExecutor if not on 2.13.
3) Using specific source files per Scala version sounds like a good plan to avoid reflection.

**Comment by patriknw** [comment]
Thanks for input

**Comment by viktorklang** [comment]
Might also want to add a comment to Akka's BatchingExecutor to NOT do any bugfixes directly there, but instead open up an Issue against scala. And just treat the code in Akka as a "shaded" class

**Comment by johanandren** [comment]
Using the execution context from the respective Scala version implemented. Doing something with the BatchingExecutionContext remains (if we can at all do anything about it)

**Comment by patriknw** [comment]
@johanandren was this fully completed or something remaining?

**Comment by johanandren** [comment]
No, only partially completed, see my comment above

**Comment by patriknw** [comment]
ok, thanks

**Comment by johanandren** [comment]
Perhaps we should be careful with backporting the BatchingExecutor given https://github.com/scala/bug/issues/12089

**Comment by johanandren** [comment]
Looking into this a little bit more the current batching impl we have has the same behavior as the 2.13.2 one, so it probably doesn't matter if we update to a faster one with the same behavior.

**Comment by mushtaq** [comment]
> Looking into this a little bit more the current batching impl we have has the same behavior as the 2.13.2 one, so it probably doesn't matter if we update to a faster one with the same behavior.

@johanandren, @patriknw  Can you please confirm if `ActorSystem[_].executionContext` in Akka 2.6.8 (with Scala 2.13.2) will have the same issue as of Scala's global ExecutionContext raised in scala/bug#12089?

**Comment by johanandren** [comment]
@mushtaq Yes it will, and has behaved like this for a very long while. See #29415 that I just created as well.

**Comment by viktorklang** [comment]
The impl of the BatchingExecutor is absolutely fine.
-- 
Cheers,
√
